File: scripts/oop-practice-sources/lua/bank-account.lua
local BankAccount = {}

function BankAccount:new(owner)
    local obj = setmetatable({owner = owner, _balance = 1000}, {__index = self})
    return obj
end

function BankAccount:deposit(amount)
    self._balance = self._balance + amount
    print("Пополнение: +" .. amount .. " ₽. Баланс: " .. self._balance .. " ₽")
end

function BankAccount:withdraw(amount)
    if amount > self._balance then
        print("Ошибка: недостаточно средств")
        return
    end
    self._balance = self._balance - amount
    print("Снятие: -" .. amount .. " ₽. Баланс: " .. self._balance .. " ₽")
end

function BankAccount:show_balance()
    print("Текущий баланс: " .. self._balance .. " ₽")
end

local account = BankAccount:new("Иван")
account:deposit(500)
account:withdraw(200)
account:show_balance()
print("Попытка прямого доступа к балансу...")
if account.balance ~= nil then
    account.balance = 999999
    print("Взлом удался!")
else
    print("Ошибка: прямой доступ к балансу запрещён")
end
account:show_balance()

File: scripts/oop-practice-sources/java/bank-account.java
class BankAccount {
    private String owner;
    private int balance;

    BankAccount(String owner) {
        this.owner = owner;
        this.balance = 1000;
    }

    void deposit(int amount) {
        balance += amount;
        System.out.println("Пополнение: +" + amount + " ₽. Баланс: " + balance + " ₽");
    }

    void withdraw(int amount) {
        if (amount > balance) {
            System.out.println("Ошибка: недостаточно средств");
            return;
        }
        balance -= amount;
        System.out.println("Снятие: -" + amount + " ₽. Баланс: " + balance + " ₽");
    }

    void showBalance() {
        System.out.println("Текущий баланс: " + balance + " ₽");
    }
}

public class Main {
    public static void main(String[] args) {
        BankAccount account = new BankAccount("Иван");
        account.deposit(500);
        account.withdraw(200);
        account.showBalance();
        System.out.println("Попытка прямого доступа к балансу...");
        // account.balance = 999999; // ошибка компиляции: balance has private access
        System.out.println("Ошибка: прямой доступ к балансу запрещён");
        account.showBalance();
    }
}
